state/interfaces_test.go

```diff
@@ -302,3 +399,361 @@ func (s *interfacesStateSuite) TestMachineInterfaceReturnsInterface(c *gc.C) {
+	s.assertNoInterfacesOnMachine(c, s.otherStateMachine)
+
+	args := []state.InterfaceArgs{{
+		Name: "foo",
+		Type: state.EthernetInterface,
+	}, {
+		Name:       "foo.42",
+		Type:       state.VLAN_8021QInterface,
+		ParentName: "foo",
+	}}
+	s.addInterfacesMultipleArgsSucceedsAndEnsureAllAdded(c, args)
+
+	results, err := s.machine.AllInterfaces()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(results, gc.HasLen, 2)
+	c.Assert(results[0].Name(), gc.Matches, "(foo|foo.42)")
+	c.Assert(results[1].Name(), gc.Matches, "(foo|foo.42)")
+
+	s.assertNoInterfacesOnMachine(c, s.otherStateMachine)
+}
+
+func (s *interfacesStateSuite) TestInterfaceRemoveFailsWithExistingChildren(c *gc.C) {
+	args := []state.InterfaceArgs{{
+		Name: "parent",
+		Type: state.BridgeInterface,
+	}, {
+		Name:       "one-child",
+		Type:       state.EthernetInterface,
+		ParentName: "parent",
+	}, {
+		Name:       "another-child",
+		Type:       state.EthernetInterface,
+		ParentName: "parent",
+	}}
+	s.addInterfacesMultipleArgsSucceedsAndEnsureAllAdded(c, args)
+
+	parent, err := s.machine.Interface("parent")
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = parent.Remove()
+	expectedError := fmt.Sprintf(
+		"cannot remove %s: parent interface %q has children: another-child, one-child",
+		parent, parent.Name(),
+	)
+	c.Assert(err, gc.ErrorMatches, expectedError)
+	c.Assert(err, jc.Satisfies, state.IsParentInterfaceHasChildrenError)
+}
+
+func (s *interfacesStateSuite) TestInterfaceRemoveSuccess(c *gc.C) {
+	existingInterface := s.addSimpleInterface(c)
+
+	s.removeInterfaceAndAssertSuccess(c, existingInterface)
+	s.assertNoInterfacesOnMachine(c, s.machine)
+}
+
+func (s *interfacesStateSuite) removeInterfaceAndAssertSuccess(c *gc.C, givenInterface *state.Interface) {
+	err := givenInterface.Remove()
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *interfacesStateSuite) TestInterfaceRemoveTwiceStillSucceeds(c *gc.C) {
+	existingInterface := s.addSimpleInterface(c)
+
+	s.removeInterfaceAndAssertSuccess(c, existingInterface)
+	s.removeInterfaceAndAssertSuccess(c, existingInterface)
+	s.assertNoInterfacesOnMachine(c, s.machine)
+}
+
+func (s *interfacesStateSuite) TestMachineRemoveAllInterfacesDeletesAllInterfaces(c *gc.C) {
+	s.assertNoInterfacesOnMachine(c, s.machine)
+
+	args := []state.InterfaceArgs{{
+		Name: "foo",
+		Type: state.EthernetInterface,
+	}, {
+		Name:       "bar",
+		Type:       state.VLAN_8021QInterface,
+		ParentName: "foo",
+	}}
+	s.addInterfacesMultipleArgsSucceedsAndEnsureAllAdded(c, args)
+
+	err := s.machine.RemoveAllInterfaces()
+	c.Assert(err, jc.ErrorIsNil)
+	s.assertNoInterfacesOnMachine(c, s.machine)
+}
+
+func (s *interfacesStateSuite) TestMachineRemoveAllInterfacesNoErrorIfNoInterfacesExist(c *gc.C) {
+	s.assertNoInterfacesOnMachine(c, s.machine)
+
+	err := s.machine.RemoveAllInterfaces()
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *interfacesStateSuite) TestInterfaceRefreshUpdatesStaleDocData(c *gc.C) {
+	fooInterface := s.addSimpleInterface(c)
```

Comment:
This test is unclear. Please add a comment telling us how fooInterface.Refresh() is able to change hardware address. It seems that fooInterface.Refresh is looking up some details by interface name.